### PR TITLE
BUG: Timestamp retains frequency of input Timestamps

### LINKED ISSUE #22311: pd.Timestamp(timestamp) looses freq
#### Code Sample, a copy-pastable example if possible

```python
In [2]: print(pd.Timestamp(pd.Timestamp('2010-08-08', freq='D')).freq)
None
```

#### Problem description

I expect a timestamp initialized from a timestamp to keep its properties. And indeeed:
```
In [3]: print(pd.Timestamp(pd.Timestamp('2010-08-08', tz='utc')).tz)
UTC
```

#### Expected Output
``'D'``

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 9123f7d0be4d6cff6cd7455b039f556c4acd2275
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-6-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.24.0.dev0+457.g9123f7d0b
pytest: 3.5.0
pip: 9.0.1
setuptools: 39.2.0
Cython: 0.28.4
numpy: 1.14.3
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.2.2.post1634.dev0+ge8120cf6d
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #22311
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
LGTM.

Do any of the other constructors lose freq?  (Eg to_datetime)

**Comment by mroeschke** [comment]
`to_datetime` looks good on master:

```
In [1]: pd.to_datetime(pd.Timestamp('2010-08-08', freq='D')).freq
Out[1]: <Day>
```

`DatetimeIndex` looks like it drops the frequency (but I am not sure if it should be kept in this case?)

```
In [3]: pd.DatetimeIndex([pd.Timestamp('2010-08-08', freq='D')]).freq
```

**Review by gfyoung — APPROVED** [review]

File: pandas/_libs/tslibs/timestamps.pyx [hunk 1]
```diff
@@ -731,8 +732,11 @@ class Timestamp(_Timestamp):
         if ts.value == NPY_NAT:
             return NaT
 
-        if is_string_object(freq):
+        if is_string_object(freq) or isinstance(freq, _BaseOffset):
```
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx]: don't do this as could have an adverse perf impact (this is a python object). use ``is_offset_object``

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/timestamps.pyx [hunk 2]
```diff
@@ -733,6 +734,9 @@ class Timestamp(_Timestamp):
 
         if is_string_object(freq):
             freq = to_offset(freq)
+        elif not is_offset_object(freq):
+            # GH 22311: Try to extract the frequency of a given Timestamp input
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/timestamps.pyx]: What do we expect it to be in this case? Aside from None I guess

Comment (mroeschke) [reply pandas/_libs/tslibs/timestamps.pyx]: This can be a `Timestamp` in which we want to extract its frequency and pass along.

Comment (jbrockmendel) [reply pandas/_libs/tslibs/timestamps.pyx]: IIUC you're referring to `ts_input` being a Timestamp.  I'm asking what `freq` is in this case.  More specifically, anything other than `None` seems like it should be invalid.  Or am I missing something?

Comment (mroeschke) [reply pandas/_libs/tslibs/timestamps.pyx]: Ah sorry for misunderstanding. You're correct in that this is essentially swallowing any nonsense frequency and just returning `None`

Comment (mroeschke) [reply pandas/_libs/tslibs/timestamps.pyx]: `freq` should only ever be `None`, a offset string, or offset object

**Comment by jreback** [comment]
lgtm. can you rebase and also do a quick perf check.

**Review by jreback — APPROVED** [review]

**Comment by mroeschke** [comment]
Looks like there's a _little_ hit (I think the asvs make it seem worse than reality), but imo not too bad 
```
$ asv continuous -f 1.1 upstream/master HEAD -b ^groupby

       before           after         ratio
     [d0691e05]       [2dcb1fbb]
+     1.29±0.04μs       2.21±0.6μs     1.71  timestamp.TimestampConstruction.time_parse_iso8601_no_tz
+     1.27±0.03μs       2.07±0.2μs     1.62  timestamp.TimestampConstruction.time_fromordinal
+     1.63±0.04μs       2.29±0.1μs     1.41  timestamp.TimestampConstruction.time_parse_today
+     1.59±0.01μs      2.22±0.06μs     1.40  timestamp.TimestampConstruction.time_parse_now
+      1.56±0.1μs       2.08±0.1μs     1.34  timestamp.TimestampConstruction.time_fromtimestamp

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.


# These two are the benchmark from timestamp.TimestampConstruction.time_parse_iso8601_no_tz
--branch
In [1]: %timeit pd.Timestamp('2017-08-25 08:16:14')
1.62 µs ± 42.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

--master
In [1]: %timeit pd.Timestamp('2017-08-25 08:16:14')
1.12 µs ± 13.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

```

File: pandas/_libs/tslibs/timestamps.pyx (hunk 2 above)
Comment (toobaz) [reply pandas/_libs/tslibs/timestamps.pyx]: Won't this silently pass if ``freq`` has invalid type? Couldn't we just say that ``freq = getattr(ts_input, 'freq', None) if freq is None else freq``?

**Comment by jreback** [comment]
thanks @mroeschke

**Pull request merged**